utf8proc.c

```diff
@@ -389,7 +412,21 @@ UTF8PROC_DLLEXPORT utf8proc_ssize_t utf8proc_decompose_char(utf8proc_int32_t uc,
     utf8proc_bool boundary;
     int tbc = property->boundclass;
     boundary = grapheme_break(*last_boundclass, tbc);
-    *last_boundclass = tbc;
+    // Special support for GB 12/13 made possible by GB999. After two RI
+    // class codepoints we want to force a break. Do this by resetting the
+    // second RI's bound class to UTF8PROC_BOUNDCLASS_OTHER, to force a break
+    // after that character according to GB999 (unless of course such a break is
+    // forbidden by a different rule such as GB9).
+    if (*last_boundclass == tbc && tbc == UTF8PROC_BOUNDCLASS_REGIONAL_INDICATOR)
+      *last_boundclass = UTF8PROC_BOUNDCLASS_OTHER;
+    // Special support for GB10. Fold any EXTEND codepoints into the previous
+    // boundclass if we're dealing with an emoji base boundclass.
+    else if ((*last_boundclass == UTF8PROC_BOUNDCLASS_E_BASE      ||
+              *last_boundclass == UTF8PROC_BOUNDCLASS_E_BASE_GAZ) &&
+             tbc == UTF8PROC_BOUNDCLASS_EXTEND)
+      *last_boundclass = UTF8PROC_BOUNDCLASS_E_BASE;
+    else
+      *last_boundclass = tbc;
```

Comment:
Would be nice to export this logic somehow so that we can use it in the Julia grapheme iterator.